### PR TITLE
Added Unique ID Functions/Updated Support Matrix

### PR DESCRIPTION
**Pull request by sommersoft**
/shared-bindings/index.rst: updated Support Matrix format as discussed in PR #503 & Issue #448.

Issue #462 :
/shared-bindings/microcontroller/Processor.x: 
  * Added functionality for retrieving the unique identifiers for all ports that support it. 
  * Sphinx-rST documentation is in place, to include a note that the nRF port does not currently include `struct` for bytearray unpacking.

/ports/XXXX/common-hal/microcontroller/Processor.x: 
  * Added functions to retreive unique identifiers for the following ports: SAMDxx, ESP8266, and nRF5x. 
  * SAMD21 and SAMD51 use different memory addresses (handled in the function), whereas the nRF51x and nRF52x use the same memory addresses. ESP8266 uses the same functionality as `machine.unique_id` (which is still available).

P.S. I apologize for the merges/etc...trying to minimize the squashes necessary. Not sure I made that mark...

### CODE REVIEW

**Comment by sommersoft** [comment]
Hold of on merging. Just compiled on my new Feather M0 Express, and I'm getting some weird results:
```

Adafruit CircuitPython 3.0.0-alpha.1-62-g111faa1-dirty on 2018-01-15; Adafruit Feather M0 Express with samd21g18
>>> import microcontroller
>>> microcontroller.cpu.uid
bytearray(b'TRJ=\x01\x00\x00\x00(a\x00 @\x13\x00 ')
>>>
```

**Comment by tannewt** [comment]
Why is that result weird?

**Comment by sommersoft** [comment]
yeah, nevermind. I was actually getting errors on the `struct.unpack`...but now it isn't. must have been a case of "error: sommersoft did something wrong". 

```

>>> import microcontroller
>>> import struct as struct
>>> print(struct.unpack('<i', microcontroller.cpu.uid))
(1028280916,)
>>>
```

File: shared-bindings/microcontroller/Processor.c [hunk 1]
```diff
@@ -93,9 +93,34 @@ const mp_obj_property_t mcu_processor_temperature_obj = {
     },
 };
 
+//| .. attribute:: unique id
+//|
+//|   Return the unique id (aka serial number) of the chip.
+//|   Returns a bytearray object. Use the CPython ``struct``
+//|   library to unpack the bytearray. Unpacked result is
+//|   a singleton (single item tuple).
+//|   _Note: ``struct`` is currently not available on the nRF port._
```
Comment (tannewt) [inline shared-bindings/microcontroller/Processor.c]: I wouldn't mention struct here. People can do what they want with it afterwards.

File: shared-bindings/microcontroller/Processor.c [hunk 2]
```diff
@@ -93,9 +93,34 @@ const mp_obj_property_t mcu_processor_temperature_obj = {
     },
 };
 
+//| .. attribute:: unique id
```
Comment (tannewt) [inline shared-bindings/microcontroller/Processor.c]: `..attribute:: uid` It should match the entry in `mcu_processor_locals_dict_table` below.

**Comment by sommersoft** [comment]
Changes made. SAMD tested on Feather M0 Express; nRF tested on Feather nRF52.

**Comment by tannewt** [comment]
Please rebase these onto latest master. Ping me on [discord](https://adafru.it/discord) if you need help.

**Pull request merged**